### PR TITLE
Prioritized Double IQN

### PR DESCRIPTION
**Pull request by prabhatnagarajan**
We should first merge Double IQN.

Here is a comparison against Double IQN

| Game        | DoubleIQN | Prioritized Double IQN |
| ------------- |:-------------:|:-------------:|
| Asterix | 507353.8|  **738166.66**|
| Bowling | **80.33**| 72.72|
| Hero | 28564.58| **35293.26**|
| MontezumaRevenge| **5.55**| 3.79|
| Qbert | 29531.1| **25763.95**|
| Seaquest | 30870.0| **31905.0**|
| Venture | 719.51|**1369.84** |
| VideoPinball| **731942.25**|717376.0|

Prioritized IQN wins on 5/7 domains and loses on 2/7!

### CODE REVIEW

File: chainerrl/agents/iqn.py [hunk 1]
```diff
@@ -223,7 +224,53 @@ def compute_eltwise_huber_quantile_loss(y, t, taus, huber_loss_threshold=1.0):
...
+        # mean over N_prime, then sum over (batch_size, N)
+        loss = F.sum(F.mean(eltwise_loss, axis=2))
+    else:
+        # mean over (batch_size, N_prime), then sum over N
+        loss = F.sum(F.mean(eltwise_loss, axis=(0, 2)))
+
+    return loss
+
+
+def compute_weighted_value_loss(eltwise_loss, batch_size, weights,
```
Comment (toslunar) [inline chainerrl/agents/iqn.py]: `batch_size` looks verbose.  How about `batch_size = eltwise_loss.shape[0]`?

File: examples/atari/train_double_iqn.py [hunk 2]
```diff
@@ -140,7 +143,15 @@ def make_env(test):
     opt = chainer.optimizers.Adam(5e-5, eps=1e-2 / args.batch_size)
     opt.setup(q_func)
 
-    rbuf = replay_buffer.ReplayBuffer(10 ** 6)
+    if args.prioritized:
+        betasteps = args.steps / args.update_interval
+        rbuf = replay_buffer.PrioritizedReplayBuffer(
+            10 ** 6, alpha=0.5, beta0=0.4, betasteps=betasteps,
```
Comment (toslunar) [inline examples/atari/train_double_iqn.py]: I didn't find `alpha` for prioritized replay in the paper https://arxiv.org/abs/1710.10044.  Could you confirm where the hyperparameters came from?

File: chainerrl/agents/iqn.py [hunk 3]
```diff
@@ -223,7 +224,53 @@ def compute_eltwise_huber_quantile_loss(y, t, taus, huber_loss_threshold=1.0):
...
+    """Compute a loss for value prediction problem.
+
+    Args:
+        eltwise_loss (Variable): Element-wise loss per example
+        batch_accumulator (str): 'mean' or 'sum'. 'mean' will use the mean of
+            the loss values in a batch. 'sum' will use the sum.
+    Returns:
+        (Variable) scalar loss
+    """
+    assert batch_accumulator in ('mean', 'sum')
```
Comment (toslunar) [inline chainerrl/agents/iqn.py]: How about `assert eltwise_loss.ndim == 3`?

File: chainerrl/agents/iqn.py [hunk 4]
```diff
@@ -223,7 +224,53 @@ def compute_eltwise_huber_quantile_loss(y, t, taus, huber_loss_threshold=1.0):
...
+    """Compute a loss for value prediction problem.
+
+    Args:
+        eltwise_loss (Variable): Element-wise loss per example
+        weights (ndarray): Weights for y, t.
+        batch_accumulator (str): 'mean' will divide loss by batchsize
+    Returns:
+        (Variable) scalar loss
+    """
+    assert batch_accumulator in ('mean', 'sum')
```
Comment (toslunar) [inline chainerrl/agents/iqn.py]: How about `assert eltwise_loss.ndim == 3`?

File: chainerrl/agents/iqn.py (hunk 1 above)
Comment (prabhatnagarajan) [reply chainerrl/agents/iqn.py]: Ok.

File: examples/atari/train_double_iqn.py (hunk 2 above)
Comment (prabhatnagarajan) [reply examples/atari/train_double_iqn.py]: The hyperparameters don't come from anywhere. Prioritized Double IQN is a new algorithm.

Comment (prabhatnagarajan) [reply examples/atari/train_double_iqn.py]: However, I could use `alpha=0.2` and `beta0=0.4`, which appears to be the case here: https://github.com/valeoai/rainbow-iqn-apex/blob/master/rainbowiqn/args.py, which implements Rainbow IQN.

File: chainerrl/agents/iqn.py (hunk 3 above)
Comment (prabhatnagarajan) [reply chainerrl/agents/iqn.py]: Ok.

File: chainerrl/agents/iqn.py (hunk 4 above)
Comment (prabhatnagarajan) [reply chainerrl/agents/iqn.py]: Ok.

File: chainerrl/agents/iqn.py [hunk 5]
```diff
@@ -5,12 +5,10 @@
 from builtins import *  # NOQA
 from future import standard_library
 standard_library.install_aliases()  # NOQA
-
 import chainer
 from chainer import cuda
 import chainer.functions as F
 import chainer.links as L
-import numpy as np
```
[the remark this answers is not in the document]
Comment (prabhatnagarajan) [reply chainerrl/agents/iqn.py]: Done.

**Pull request merged**